gcs/src/main/scala/org/broadinstitute/monster/storage/gcs/GcsApi.scala

```diff
@@ -86,6 +98,261 @@ class GcsApi private[gcs] (runHttp: Request[IO] => Resource[IO, Response[IO]]) {
+
+    val dataHeader =
+      s"""--${multipartBoundary.value}
+         |${`Content-Type`(MediaType.application.json, Charset.`UTF-8`).renderString}
+         |
+         |${objectMetadata.noSpaces}
+         |
+         |--${multipartBoundary.value}
+         |${contentType.renderString}
+         |""".stripMargin.getBytes
+
+    val dataFooter =
+      s"""
+         |--${multipartBoundary.value}--""".stripMargin.getBytes
+
+    val fullBody = Stream
+      .emits(dataHeader)
+      .append(data)
+      .append(Stream.emits(dataFooter))
+
+    fullBody.compile.toChunk.flatMap { chunk =>
+      val fullHeaders = Headers.of(
+        `Content-Type`(MediaType.multipartType("related", Some(multipartBoundary.value))),
+        `Content-Length`.unsafeFromLong(chunk.size.toLong)
+      )
+
+      val gcsReq = Request[IO](
+        method = Method.POST,
+        uri = baseGcsUploadUri(bucket, "multipart"),
+        headers = fullHeaders,
+        body = Stream.chunk(chunk)
+      )
+      runHttp(gcsReq).use { response =>
+        if (response.status.isSuccess) {
+          IO.unit
+        } else {
+          IO.raiseError(
+            new RuntimeException(s"Failed to upload bytes to $path in $bucket")
+          )
+        }
+      }
+    }
+  }
+
+  /**
+    * Initialize a GCS resumable upload.
+    *
+    * Resumable uploads enable pushing data to an object in multiple chunks, and are recommended
+    * for use with files larger than 5MB. The initialization request for the upload must include
+    * any metadata for the object that will be created.
+    *
+    * @see https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload#start-resumable
+    *
+    * @param bucket the GCS bucket to create the new object within
+    * @param path path within `bucket` where the new object will be created
+    * @param contentType content-type to set on the new object
+    * @param expectedSize total number of bytes that will be uploaded to the new object over
+    *                     the course of all subsequent requests
+    * @param expectedMd5 expected md5, if any, of the new object. Setting this will enable
+    *                    server-side content validation in GCS
+    * @return the unique ID of this upload, for use in subsequent requests
+    */
+  def initResumableUpload(
+    bucket: String,
+    path: String,
+    contentType: `Content-Type`,
+    expectedSize: Long,
+    expectedMd5: Option[String]
+  ): IO[String] = {
+    val objectMetadata = buildUploadMetadata(path, expectedMd5).noSpaces.getBytes
+
+    val gcsReq = Request[IO](
+      method = Method.POST,
+      uri = baseGcsUploadUri(bucket, "resumable"),
+      body = Stream.emits(objectMetadata),
+      headers = Headers.of(
+        `Content-Length`.unsafeFromLong(objectMetadata.length.toLong),
+        `Content-Type`(MediaType.application.json, Charset.`UTF-8`),
+        Header("X-Upload-Content-Length", expectedSize.toString),
```

Comment:
Can you pull `X-Upload-Content-Length` and `X-Upload-Content-Type` into constants on the `GcsApi` object?